### PR TITLE
Enable support for gitlab

### PR DESCRIPTION
**Pull request by ipa-mdl**
This PR includes #119 

Example: https://gitlab.com/ipa-mdl/industrial_ci/pipelines/6437124

gitlab.com runs the CI tests on shared Docker-based runners for free.
This means that Docker needs to be run in Docker..
The tricky part was the prerelease test: Docker is run by Docker, which is run in Docker.

The `_DO_NOT_FOLD` option should be migrated into the `STYLE` feature, it is just a preliminary hack.

@VictorLamoine, @ipa-bnm: FYI

### CODE REVIEW

**Comment by shaun-edwards** [comment]
Really excited to see this feature added. Looks like a good number of changes were required. Thanks @ipa-mdl for the effort.

**Comment by ipa-mdl** [comment]
> Looks like a good number of changes were required

most changes were introduced in https://github.com/ros-industrial/industrial_ci/pull/120/commits/3a9a91006549486773a953d8266643a5ecedf1d1, which was not necessary for gitlab support. I was just another clean-up step before copying things..
Only the last 2 commits address gitlab support directly.

**Comment by ipa-mdl** [comment]
v0.3.1 executed `BEFORE_SCRIPT` in the src space, the master executes it in `$TARGET_REPO_PATH` (which ir read-only) and this patch will run it in `/` (default docker workdir),

Should we enforce a default working directory in the docker container? Ans which one?

**Comment by 130s** [comment]
Sorry I don't have much time these days to look into a huge change like this. And I don't have working Gitlab setting right now to test. 

Travis seems happy with this change, which is a sign that we can merge as long as someone can approve :). 
It'd be ideal if we can integrate a script to setup another OS where Gitlab is installed and running, so that we can continuously integrate `industrial_ci` from Gitlab (I'm not pushing this idea for this PR though).

> Should we enforce a default working directory in the docker container? 

As long as the existing users' script used in `BEFORE_SCRIPT` work, I'm fine I think.

**Comment by ipa-mdl** [comment]
I have set-up a mirror of my fork on gitlab.com, which will get tested.

**Comment by 130s** [comment]
During today's dev mtg, @shaun-edwards offered to test this on his GitLab environment.

**Comment by ipa-mdl** [comment]
@shaun-edwards: if you don't user Docker base-runners, you have to prepare docker within the `before_script. The rest should work out-of-the-box.

**Comment by ipa-bnm** [comment]
This feature comes in very handy. I will also test it on our gitlab environment. Thanks @ipa-mdl

**Comment by ipa-mdl** [comment]
I have rebase it to the latest master.
Gitlab still works: https://gitlab.com/ipa-mdl/industrial_ci/pipelines/6957440

**Comment by ipa-mdl** [comment]
I have sucessfully tested it wih gitlab.com and a local gitlab instance.
@shaun-edwards, @VictorLamoine: do you have time to test this within your set-up?

**Comment by VictorLamoine** [comment]
No sorry :dizzy_face: I will, but I don't know when

File: gitlab.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+# Copyright (c) 2016, Isaac I. Y. Saito
```
Comment (130s) [inline gitlab.sh]: Copyright must be @ipa-mdl ;)

**Comment by shaun-edwards** [comment]
@ipa-mdl, looking at this right now.  I'll let you know if I have any questions/issues.

**Comment by shaun-edwards** [comment]
I wasn't able to get it to work, but I'm sure it's just user error.  I mirrored `industrial_core` for testing purposes.  The test fails [here](https://gitlab.com/shaun-edwards/industrial_core_mirror/builds/12612234).  It appear my issue is trying to clone the `industrial_ci` repo.  The script fails with this message:
```
$ git clone -b feature/gitlab https://github.com/ipa-mdl/industrial_ci
/bin/sh: eval: line 58: git: not found
```
My confusion is caused by a lack of an `install` tag in the gilab yaml format.  Normally I would just add these [lines](https://github.com/ros-industrial/industrial_core/blob/kinetic-devel/.travis.yml#L19-L20) to the Travis yaml file.

**Comment by ipa-mdl** [comment]
The gitlab docker-in-docker feature uses the alpine-based docker image.
Please try to add git as an additional package: https://github.com/ipa-mdl/industrial_ci/blob/feature/gitlab/.gitlab-ci.yml#L8

**Comment by shaun-edwards** [comment]
I think I did that correctly [here](https://gitlab.com/shaun-edwards/industrial_core_mirror/blob/kinetic-devel/.gitlab-ci.yml).

**Comment by ipa-mdl** [comment]
just change https://gitlab.com/shaun-edwards/industrial_core_mirror/blob/kinetic-devel/.gitlab-ci.yml#L2 to `image: docker:git`.
It contains docker and git :)

**Comment by ipa-mdl** [comment]
` git clone -b feature/gitlab https://github.com/ipa-mdl/industrial_ci .industrial_ci`

**Comment by shaun-edwards** [comment]
This change is working for me.  I suggest we add some instruction to the README or advanced instructions for setting up Gitlab.  It's basically a file similar to the `travis.yml` but with a little bit different syntax.  Here is a good [example](https://gitlab.com/shaun-edwards/industrial_core_mirror/blob/kinetic-devel/.gitlab-ci.yml) - Note the `git clone` command will be updated once this PR is accepted.

Thanks for all the hard work @ipa-mdl!

**Comment by ipa-mdl** [comment]
@shaun-edwards: Thanks for your test!
I have updated the example config, it now tests `industrial_ci` with the upstream `industrial_ci`.

The current environment variable passing is based on the matrix feature of Travis CI.
For Gitlab CI each job gets a script, so an alternate syntax with arguments passing is possible (same as for   as for `run_ci`. For the sake of completeness, `travis.sh` supports it as well.

**Comment by ipa-mdl** [comment]
@shaun-edwards: I cannot use my Gitlabs jobs right now, because of the git clone from master. Can you restart your pipeline to test my latest fixes?

**Comment by shaun-edwards** [comment]
Looks like everything is still working: https://gitlab.com/shaun-edwards/industrial_core_mirror/pipelines/7159777

**Comment by ipa-mdl** [comment]
Thanks again!
The major difference is that the colored `>>>` and `<<<` are included again.
The verbose rosinstall output git limited in #152 

I will merge it tomorrow if @130s is okay with it.

**Pull request merged**